### PR TITLE
UART is now stable.

### PR DESCRIPTION
**Pull request by PyroAVR**
+ Documentation updated in files and headers
 + Submodules added for queue library

 - Removed stale code and files from previous asm driver
 - Removed unused kernel interfaces sbrk, abort
 - Renamed kernel_support to file_table

### CODE REVIEW

File: src/drivers/kinetis/uart.c [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+queue_t rxq_headers[6];
+volatile char txqs[6][255];
+volatile char rxqs[6][255];
+uart_context contexts[6] = {
+    {UART0, &txq_headers[0], &rxq_headers[0]},
+    {UART1, &txq_headers[1], &rxq_headers[1]},
+    {UART2, &txq_headers[2], &rxq_headers[2]},
+    {UART3, &txq_headers[3], &rxq_headers[3]},
+    {UART4, &txq_headers[4], &rxq_headers[4]},
+    {UART5, &txq_headers[5], &rxq_headers[5]}
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c:57]: we don't have access to UART5

File: src/drivers/kinetis/uart.c [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    isr_active = false;
+
+    //Enable transmitter and receiver of UART (and interrupts)
+    if(conf.configure_interrupts > 0) {
+        asm("cpsid i");
+        // set RWFIFO[RXWATER] such that interrupts are actually generated
+        conf.uart_base->RWFIFO = 1;
+        // set TWFIFO[TXWATER] such that interrupts are actually generated
+        /*
+         *conf.uart_base->TWFIFO = 1;
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c]: stray comments o_o

File: src/drivers/kinetis/uart.c [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+ * Interrupt Service Routines
+ *     state and prototypes
+ */
+/**
+ * Generic ISR for UART RX/TX interrupts
+ */
+void uart_isr(void);
+
+// currently active uart
+static volatile int which_uart;
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c]: cant this global be avoided if the UART number is passed into the generic ISR

File: src/drivers/kinetis/uart.c [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+// true if any uart is currently being processed by uart_isr
+static volatile bool isr_active;
+
+queue_t txq_headers[6];
+queue_t rxq_headers[6];
+volatile char txqs[6][255];
+volatile char rxqs[6][255];
+uart_context contexts[6] = {
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c:51]: does every queue get created even if we don't use a specific UART?

File: src/drivers/kinetis/uart.c [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+ */
+/**
+ * Generic ISR for UART RX/TX interrupts
+ */
+void uart_isr(void);
+
+// currently active uart
+static volatile int which_uart;
+// true if any uart is currently being processed by uart_isr
+static volatile bool isr_active;
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c]: whats going on with this? do we need to lock per iqr_handler?

File: src/drivers/kinetis/uart.c (hunk 1 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c:57]: generic driver for K64

File: src/drivers/kinetis/uart.c (hunk 2 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c]: these are not stray - this is something that we can optionally enable.  Do you want it enable-able by the config struct?

File: src/drivers/kinetis/uart.c (hunk 4 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c:51]: yes, they have to be statically allocated, as we do not have dynamic memory in the kernel yet.  When we have that, these can go.

File: src/drivers/kinetis/uart.c (hunk 5 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c]: probs not, time to find out if it breaks

File: src/drivers/kinetis/uart.c (hunk 2 above)
Comment (spressomartini) [reply src/drivers/kinetis/uart.c]: probably a good idea

Comment (PyroAVR) [reply src/drivers/kinetis/uart.c]: Fixed now - these are configurable per-uart.

File: src/drivers/kinetis/uart.c (hunk 3 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c]: Fixed now - constant is passed in at the entry point functions.

File: src/drivers/kinetis/uart.c (hunk 5 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c]: Apparently not, I removed this.

**Comment by PyroAVR** [comment]
test branch is `test/uart_subsystem`.
Pull from that branch and run on the FRDM board.  The LED will flicker between red and blue if there is significant latency in interrupt handling.  It will appear purple if the system is running smoothly.
Torture test: `$ yes > /dev/k64_serial_port_name`

The system will start to stutter under the torture test without FIFOs enabled, but once they are turned on, the operation returns to normal (no flickering).

File: src/drivers/kinetis/uart.c [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    int which = (conf.uart_base == UART0)? 0:
+                (conf.uart_base == UART1)? 1:
+                (conf.uart_base == UART2)? 2:
+                (conf.uart_base == UART3)? 3:
+                (conf.uart_base == UART4)? 4:5;
+
+    queue_init(&txq_headers[which], &txqs[which], 255);
+    queue_init(&rxq_headers[which], &rxqs[which], 255);
+
+    // make isr available
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c:119]: whats going on with `which_uart` and `isr_active` here?

File: src/drivers/kinetis/uart.c [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    return bytes_read;
+}
+
+
+unsigned int uart_close(uart_context *context) { return 0; }
+
+
+void uart_isr(int which_uart){
+    uint8_t sz = 0;
+    asm("cpsid i");
```
Comment (spressomartini) [inline src/drivers/kinetis/uart.c:211]: whats the purpose of turning off interrupts here?

Comment (PyroAVR) [reply src/drivers/kinetis/uart.c:211]: No interruption during the ISR - if we allow interruption, the data may be lost (this was likely one of the causes of bugs in the asm uart driver)

File: src/drivers/kinetis/uart.c (hunk 6 above)
Comment (PyroAVR) [reply src/drivers/kinetis/uart.c:119]: e r g, I'll fix it

**Review by spressomartini — APPROVED** [review]
Looks good to me! 
Remember to make note of the blocking compiler flag, the compile submodules statically flag, and if no blocking is used, use higher interrupt priority for the UART.

**Comment by PyroAVR** [comment]
Those changes will be made in a future pr to update documentation, as well as some quality of life updates (like configurable buffers).  Merging for now so that we can move on with development until this is critical again.

**Pull request merged**